src/uharfbuzz/_harfbuzz.pyx

```diff
@@ -1369,3 +1394,368 @@ def repack_with_tag(tag: str,
     return packed
 
 
+def subset_preprocess(face: Face) -> Face:
+    new_face = hb_subset_preprocess(face._hb_face)
+    return Face.from_ptr(new_face)
+
+
+class SubsetInputSets(IntEnum):
+    GLYPH_INDEX = HB_SUBSET_SETS_GLYPH_INDEX
+    UNICODE = HB_SUBSET_SETS_UNICODE
+    NO_SUBSET_TABLE_TAG = HB_SUBSET_SETS_NO_SUBSET_TABLE_TAG
+    DROP_TABLE_TAG = HB_SUBSET_SETS_DROP_TABLE_TAG
+    NAME_ID = HB_SUBSET_SETS_NAME_ID
+    NAME_LANG_ID = HB_SUBSET_SETS_NAME_LANG_ID
+    LAYOUT_FEATURE_TAG = HB_SUBSET_SETS_LAYOUT_FEATURE_TAG
+    LAYOUT_SCRIPT_TAG = HB_SUBSET_SETS_LAYOUT_SCRIPT_TAG
+
+
+cdef class SubsetInput:
+    cdef hb_subset_input_t* _input
+
+    def __cinit__(self):
+        self._input = hb_subset_input_create_or_fail()
+        if self._input is NULL:
+            raise MemoryError()
+
+    def __dealloc__(self):
+        if self._input is not NULL:
+            hb_subset_input_destroy(self._input)
+
+    def subset(self, source: Face) -> Face:
+        new_face = hb_subset_or_fail(source._hb_face, self._input)
+        if new_face == NULL:
+            raise RuntimeError("Subsetting failed")
+        return Face.from_ptr(new_face)
+
+    def keep_everything(self):
+        hb_subset_input_keep_everything(self._input)
+
+    def pin_axis_to_default(self, face: Face, tag: str) -> bool:
+        hb_tag = hb_tag_from_string(tag.encode("ascii"), -1)
+        return hb_subset_input_pin_axis_to_default(
+            self._input, face._hb_face, hb_tag
+        )
+
+    def pin_axis_location(self, face: Face, tag: str, value: float) -> bool:
+        hb_tag = hb_tag_from_string(tag.encode("ascii"), -1)
+        return hb_subset_input_pin_axis_location(
+            self._input, face._hb_face, hb_tag, value
+        )
+
+    @property
+    def unicode_set(self) -> Set:
+        return Set.from_ptr(hb_set_reference (hb_subset_input_unicode_set(self._input)))
+
+    @property
+    def glyph_set(self) -> Set:
+        return Set.from_ptr(hb_set_reference (hb_subset_input_glyph_set(self._input)))
+
+    def set(self, set_type : SubsetInputSets) -> Set:
```

Comment:
I think `set` is confusing here, I’d expect a method that allows me to set stuff. What if this became a dict-like property, e.g. `name_ids_set = input.sets[SubsetInputSets.NAME_ID]` or something like that, or even `name_ids_set = input[SubsetInputSets.NAME_ID]` and using a custom `__getitem__` w/o `__setitem__`?

Reply:
I like the __getitem__ idea.  __setitem__ can also be implemented by overriding that entire set.  Let me implement that.

Reply:
I'll keep the name `sets` internally.

Reply:
Done.

Reply:
@madig had originally added properties for each of them individually. That's also an option; I wasn't a fan of it...

Reply:
Yes, I saw it. I think the current solution is better as it requires less maintenance going forward.

Reply:
But I think If we really want to do this, we can use `__getattr__`/`__setattr__`, so that something like `input.name_id` translates to `input[SubsetInputSets.NAME_ID]`.

Reply:
or `input.name_id_set`

Reply:
Yeah I thought about that too. That sounds like too much magic. I'm happy with current solution I think; either this or hand-written properties. getattr/setattr goes too far for me for most cases.

Reply:
Maybe we go back to what madig had?

Reply:
I mean, we keep the sets() or set() API, but also add properties? In HB API we specifically didn't add those. Again, I think I prefer the getitem/setitem and call it a day.

Reply:
Agree, it is good the way it is now.

Reply:
Done.